src/btree/bt_read.c

```diff
@@ -125,17 +128,258 @@ __instantiate_birthmarks(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    struct las_page_prepared_updates {
+        WT_ITEM key;
+        wt_timestamp_t timestamp;
+        uint64_t txnid;
+    } * las_preparep;
+    WT_BIRTHMARK_DETAILS *birthmarkp;
+    WT_CACHE *cache;
+    WT_CURSOR *cursor;
+    WT_CURSOR_BTREE cbt;
+    WT_DECL_ITEM(las_prepares);
+    WT_DECL_RET;
+    WT_ITEM las_key, las_value, next_las_key;
+    WT_PAGE *page;
+    WT_UPDATE *upd;
+    wt_timestamp_t durable_timestamp, las_timestamp;
+    size_t incr;
+    uint64_t birthmark_cnt, las_txnid, recno;
+    uint32_t i, las_btree_id, las_prepare_cnt, session_flags;
+    uint8_t prepare_state, upd_type;
+    const uint8_t *p;
+    int cmp, exact, instantiated_cnt;
+    bool birthmark_record, first_scan, locked;
+
+    birthmarkp = NULL;
+    cursor = NULL;
+    page = ref->page;
+    upd = NULL;
+    birthmark_cnt = 0;
+    las_btree_id = S2BT(session)->id;
+    las_prepare_cnt = 0;
+    locked = false;
+    instantiated_cnt = 0;
+    birthmark_record = false;
+    first_scan = true;
+    recno = WT_RECNO_OOB;
+    session_flags = 0; /* [-Werror=maybe-uninitialized] */
+    WT_CLEAR(las_key);
+
+    cache = S2C(session)->cache;
+
+    /*
+     * Check whether the disk image contains all the newest versions of the page or not?
+     */
+    if (ref->page_las->min_skipped_ts == WT_TS_MAX)
+        return __instantiate_birthmarks(session, ref);
+
+    __wt_btcur_init(session, &cbt);
+    __wt_btcur_open(&cbt);
+
+    WT_ERR(__wt_scr_alloc(session, 0, &las_prepares));
+    WT_STAT_CONN_INCR(session, cache_read_lookaside_skew_old);
+    WT_STAT_DATA_INCR(session, cache_read_lookaside_skew_old);
+    if (WT_SESSION_IS_CHECKPOINT(session)) {
+        WT_STAT_CONN_INCR(session, cache_read_lookaside_skew_old_checkpoint);
+        WT_STAT_DATA_INCR(session, cache_read_lookaside_skew_old_checkpoint);
+    }
+
+    /* Open a lookaside table cursor. */
+    __wt_las_cursor(session, &cursor, &session_flags);
+
+    /*
+     * The lookaside records are in key and update order, that is, there will be a set of in-order
+     * updates for a key, then another set of in-order updates for a subsequent key. We find the
+     * most recent of the updates for a key and then insert that update into the page, then all the
+     * updates for the next key, and so on. In case if the birthmark record exists for that key,
+     * then insert birthmark record into the page.
+     *
+     * The search starts with minimum LAS key of the page until the maximum LAS key (both included).
+     * As we are going to instantiate only one record for a key and that too is a most recent one.
+     * To find out the most recent record for a key, we use the search_near cursor function with
+     * both maximum timestamp and transaction id as part of the key search. This search can return
+     * either the last record of the search key or the first record of the next key.
+     *
+     * Once we instantiate the most recent record for a key, the search continues to the next key
```

Comment:
Can this be edited to include something like: "since we're pointing at the max timestamp and transaction id pairing for a given key, the next entry is guaranteed to be for the next key". Took me a moment to figure out what was going on here.

Reply:
Added those lines to make it more clear.